Python/Product/VSInterpreters/CPythonInterpreterFactoryProvider.cs

```diff
@@ -230,77 +221,90 @@ class CPythonInterpreterFactoryProvider : IPythonInterpreterFactoryProvider {
-                        }
-                    }
-                }
+            HashSet<string> registeredPaths = new HashSet<string>();
+            var arch = Environment.Is64BitOperatingSystem ? null : (ProcessorArchitecture?)ProcessorArchitecture.X86;
 
-                using (var baseKey = RegistryKey.OpenBaseKey(RegistryHive.LocalMachine, RegistryView.Registry32))
+            Dictionary<string, InterpreterInformation> userFactories = new Dictionary<string, InterpreterInformation>();
+            using (var baseKey = RegistryKey.OpenBaseKey(RegistryHive.CurrentUser, RegistryView.Default)) {
                 using (var python = baseKey.OpenSubKey(PythonPath)) {
                     if (python != null) {
-                        anyChanged |= RegisterInterpreters(registeredPaths, registeredIds, false, python, ProcessorArchitecture.X86);
+                        RegisterInterpreters(userFactories, registeredPaths, true, python, arch);
                     }
                 }
+            }
 
-                if (Environment.Is64BitOperatingSystem) {
-                    using (var baseKey = RegistryKey.OpenBaseKey(RegistryHive.LocalMachine, RegistryView.Registry64))
-                    using (var python64 = baseKey.OpenSubKey(PythonPath)) {
-                        if (python64 != null) {
-                            anyChanged |= RegisterInterpreters(registeredPaths, registeredIds, false, python64, ProcessorArchitecture.Amd64);
-                        }
+            Dictionary<string, InterpreterInformation> machineFactories = new Dictionary<string, InterpreterInformation>();
+            using (var baseKey = RegistryKey.OpenBaseKey(RegistryHive.LocalMachine, RegistryView.Registry32))
+            using (var python = baseKey.OpenSubKey(PythonPath)) {
+                if (python != null) {
+                    RegisterInterpreters(machineFactories, registeredPaths, false, python, ProcessorArchitecture.X86);
+                }
+            }
+
+            if (Environment.Is64BitOperatingSystem) {
+                using (var baseKey = RegistryKey.OpenBaseKey(RegistryHive.LocalMachine, RegistryView.Registry64))
+                using (var python64 = baseKey.OpenSubKey(PythonPath)) {
+                    if (python64 != null) {
+                        RegisterInterpreters(machineFactories, registeredPaths, false, python64, ProcessorArchitecture.Amd64);
+                    }
+                }
+            }
+
+            // Then update our cached state with the lock held.
+            lock (this) {
+                var uniqueIds = userFactories.Select(x => x.Key)
+                       .Concat(machineFactories.Select(x => x.Key))
+                       .Distinct();
+
+                foreach (var id in uniqueIds) {
+                    InterpreterInformation newInfo;
+                    // User factories with the same ID take precedence over machine factories
```

Comment:
The difficultly there is actually with the IDs themselves...  If we want to do that we need to have that show up in the interpreter ID, so that we'd end up with something like:
Global|PythonCore|3.5|x86|User
 or
Global|PythonCore|3.5|x86|Machine

Or something along those lines.  My current thinking is that will cause more trouble in cross machine scenarios then it's worth enabling users to have both a per-user and per-machine install.